Fixed Quest List Bugs: Duplicate IDs, Border Styling, Infinite Loading Indicator

Enprogames opened the pull request:
## What?
There were still a few major issues with the bootstrap-table quest lists that needed to be worked out. These included:

- **Duplicate IDs**: When a quest was clicked and the detailed content was shown, the content was duplicated from another source. Since the HTML DOM elements contained IDs (which are supposed to be kept unique), the JavaScript code wasn’t sure which elements to target. As a result, it would simply target elements that were closer to the top of the tree. This resulted in several issues, including the loading indicator being shown forever.
- **Reduce JavaScript Includes**: Some JavaScript includes (bootstrap-table.min.js and bootstrap-table-accordion.js) were done multiple times in multiple files, so they were moved to base.html.
- **Bad Table Styling**: Some table styles related to borders were improved.

## Why?
These issues would reduce user satisfaction and make the project more difficult to maintain in the long term.

## How?

### Expand/Collapse JavaScript Refactoring
Perhaps the biggest change was to the bootstrap-table-accordion.js, which is responsible for handling expanding/collapsing of quest list rows when the user clicks them. Previously, when content was being inserted underneath a quest row, it was duplicated from a hidden DIV element that existed elsewhere on the page. The hidden DIV was meant to hold the content while the row was collapsed, allowing for the ajax_content_loading.html script to load content while a row was still collapsed.

This duplication of HTML code resulted in unexpected issues, such as the element IDs existing in more than one place. This created several problems, such as the loading indicator being displayed forever if a row was expanded before it had its content inserted, then had its content inserted before being collapsed/expanded again. The reason was that the ajax_content_loading.html script would target the duplicated element. It would insert the content into the duplicated code that was inserted underneath the row, rather than in the original hidden DIV element.

The solution to this was to "move" the hidden DIV underneath the row, rather than duplicate it. This greatly simplified the expanding/collapsing process. Not implementing it this way in the first place was a definite oversight on my part.

### Table Border Styling Fixed for Muted and Collapsed Rows, and Table in Mobile View (#1403 and #1409)
I made some changes to the bootstrap-table-accordion CSS. Firstly, the border around the table in the mobile view is removed. Second, the border is made invisible while a table row is collapsing. And third, the border was removed around the "muted" expanded rows.

```css
/* Fixes #1409 */
.accordian-table {
  border: transparent;
}
.accordian-table > tbody > tr,
.accordian-table > tbody > tr > .collapsing {
  --accordian-table-border-color: transparent !important;
}
/* Fixes #1403 */
.accordian-table > tbody > tr > .muted {
  --accordian-table-border-color: transparent;
}
```

### JavaScript Includes Moved to `quest_manager/base.html`
Since several of the templates in the quest_manager app make use of the bootstrap-table library, I moved the script includes to `quest_manager/base.html`:
```html
{% block js %}
    {% comment %} ### Import Bootstrap Table JS ### {% endcomment %}
    <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-table/1.18.3/bootstrap-table.min.js"
            integrity="sha384-H2wFlf4pp5vxYPba0+XWp9zlTzEKO5Syk7ZZlbgOgZqj5GWvazhyWvq9zHs0INqj"
            crossorigin="anonymous"></script>

    <script src="{% static 'quest_manager/js/bootstrap-table-accordion.js' %}"></script>
{% endblock %}
```

Since other scripts included the `ajax_content_loading.html` script in their JS blocks, I had to use the `{{ block.super }}` tag in some of the templates, such as here in `quest.html`:
```
{% block js %}
  {% comment %} Include from quest_manager/base.html {% endcomment %}
  {{ block.super }}

  {% with quests_available_page=True %}
    {% include "quest_manager/ajax_content_loading.html" %}
  {% endwith %}

{% endblock %}
```

## Testing?
I spent time clicking through the various options that were previously broken, such as the quick reply form, the "Please read submission instruction more carefully..." button, the flag button, etc.

It would also be helpful if the reviewer checked out these changes to make sure nothing is broken.

## Screenshots (if front end is affected)

<details>

<summary>Top Border Fixed in Mobile View</summary>

![image](https://github.com/bytedeck/bytedeck/assets/11304586/f698615d-dc19-4499-a13c-a15c162b492b)

</details>

<details>

<summary>Expanded content for muted rows is no longer muted</summary>

![image](https://github.com/bytedeck/bytedeck/assets/11304586/63ef1d45-fc7f-4f88-a66b-1661366421e0)

![image](https://github.com/bytedeck/bytedeck/assets/11304586/bfa45c39-6727-4215-a86b-10f1e92f6e6e)


</details>

<details>

<summary>While a row is collapsing, the border is removed</summary>

![image](https://github.com/bytedeck/bytedeck/assets/11304586/3ad1f3cb-d407-45ab-9fa8-07c985d0e0b3)


</details>

## Anything Else?
I looked into fixing #1407, but I'm not sure if it is still issue. I'm also not entirely sure of the intended behavior. Based on the code and comments in `quest_manager/ajax_content_loading.html`, it looks like the hidden labels are only meant to show up in the quests available tab? But some comments seem to indicate that they should also show up in the quest approvals tab, They do show up in the quests available tab for me, but not in the approvals tab. I may need a bit longer to investigate this.

## Review request
@tylerecouture

tylerecouture reviewed (changes requested):
Seem to be some border inconsistancies:
![InconsistantBorders](https://github.com/bytedeck/bytedeck/assets/10604391/0f092243-76cb-44ab-91ed-e094eb60bc35)

And:
![image](https://github.com/bytedeck/bytedeck/assets/10604391/df293dc7-9b7e-4d2d-8aab-febe795ce984)

Enprogames commented:
> Seem to be some border inconsistancies: ![InconsistantBorders](https://user-images.githubusercontent.com/10604391/248151971-0f092243-76cb-44ab-91ed-e094eb60bc35.gif)
> 
> And: ![image](https://user-images.githubusercontent.com/10604391/248152016-df293dc7-9b7e-4d2d-8aab-febe795ce984.png)

That's strange. I wasn't able to replicate this. I removed some styles that might have been causing issues with this though.

tylerecouture commented:
"Finally, I added a timestamp span element to the submitted_status.html snippet:"  Those are going to need a code comment because there is no context for why they are there and will easily get deleted in some future code cleanup.

tylerecouture commented:
"I removed some styles that might have been causing issues with this though."  Were they causing the problem or not?  Do you know why those styles were there?  Detail view doesn't sound like the accordion view, it sounds like the quest detail view. Are you sure those styles weren't for the detail view?

Enprogames commented:
> "I removed some styles that might have been causing issues with this though." Were they causing the problem or not? Do you know why those styles were there? Detail view doesn't sound like the accordion view, it sounds like the quest detail view. Are you sure those styles weren't for the detail view?

From what I've been able to uncover with the git blame, it looks like most of those styles are from 6 years ago when you were experimenting with some bootstrap table accordion stuff. I'm not sure what you mean by the accordion view and quest detail view. Just so we're on the same page, when you click a row, it creates a `<tr class="detail-view">` element, and the content that is inserted has several list-group-item elements. 
![image](https://github.com/bytedeck/bytedeck/assets/11304586/3c6e6e19-e6e6-4303-add0-038845548bb2)

It looks like the specific commit that you added this in is "89dc604: accordian table scrolling", but from what I can see it doesn't look like what you were working on (`suggestions/templates/suggestions/suggestion_list.html`) is being used anymore. And before I moved those styles, I checked with a few of the other classes you defined, and they weren't in use at all anymore. 

![image](https://github.com/bytedeck/bytedeck/assets/11304586/518f1290-2fec-491a-8ea5-5b4bfa78a82e)

Enprogames commented:
> "Finally, I added a timestamp span element to the submitted_status.html snippet:" Those are going to need a code comment because there is no context for why they are there and will easily get deleted in some future code cleanup.

You're right. I meant to do that. They're there now.

tylerecouture commented on src/quest_manager/static/quest_manager/js/bootstrap-table-accordion.js:
```diff
@@ -1,3 +1,27 @@
+  if (timestampA > timestampB) return 1;
+  return 0;
+}
+
```
Uneccessary.  Just use a column with `data-visible='false'`with a numeric value for the datetime (see last_sub_datetime column in profile_list.html) and sort based on that column via `data-sort-name='name_of_other_column'`

This info was in the original issue.

Enprogames replied:
This was changed as per the description in the original issue. I also moved it to a new pull request:
https://github.com/bytedeck/bytedeck/pull/1420

tylerecouture commented on src/quest_manager/templates/quest_manager/snippets/submitted_status.html:
Revert this file.  No changes needed here.

tylerecouture commented on src/quest_manager/templates/quest_manager/tab_quests_approvals.html:
```diff
@@ -23,14 +23,14 @@
         <th class="col-sm-1 hidden-xs" data-field="group_name" data-sortable="true">{% group_name %}</th>
         <th class="col-sm-2 col-xs-3 hidden-xs" data-field="user" data-sortable="true">User</th>
-        <th class="col-sm-3 hidden-xs" data-field="status" data-sortable="true">Status</th>
+        <th class="col-sm-3 hidden-xs" data-field="status" data-sortable="true" data-sorter="dateSorter">Status</th>
```
See original issue for how to do this with only a couple lines of code using `data-sort-name`

tylerecouture commented on src/quest_manager/templates/quest_manager/tab_quests_submission.html:
```diff
@@ -21,15 +21,15 @@
         <th class="col-sm-2 text-right hidden-xs" data-field="campaign" data-sortable="true">Campaign</th>
         <th class="col-sm-1 text-center hidden-xs" data-field="tags" data-sortable="true">{{ config.custom_name_for_tag }}s</th>
-        <th class="col-sm-3 hidden-xs" data-field="status" data-sortable="true">Status</th>
+        <th class="col-sm-3 hidden-xs" data-field="status" data-sortable="true" data-sorter="dateSorter">Status</th>
```
per above

tylerecouture reviewed (changes requested):
Please submit the data sort changes as it's own PR using the method provided in the issue.
